### PR TITLE
callmemaybe refactoring

### PR DESCRIPTION
**Pull request by lubennikovaav**
- Don't spawn a separate thread for each connection.
Instead use one thread per safekeeper, that iterates over all connections and sends callback requests for them.

-Use tokio postgres to connect to the pageserver, to avoid spawning a new thread for each connection.

### CODE REVIEW

**Comment by kelvich** [comment]
Mpsc changed to crossbeam due to
https://doc.rust-lang.org/nightly/std/sync/mpsc/struct.Receiver.html#known-issues

**Comment by funbringer** [comment]
Thanks for the patch! I'll look into it tomorrow.

File: walkeeper/src/receive_wal.rs [hunk 1]
```diff
@@ -127,19 +88,27 @@ impl<'pg> ReceiveWalConn<'pg> {
...
-            let _ = thread::Builder::new()
-                .name("request_callback thread".into())
-                .spawn(move || {
-                    request_callback(conf, timelineid, tenant_id);
-                })
-                .unwrap();
-        }
+        // as long as this wal_stream is alive, callmemaybe thread
+        // will send requests to pageserver
+        let _wal_stream = match swh.conf.pageserver_addr {
```
Comment (LizardWizzard) [inline walkeeper/src/receive_wal.rs:93]: If I understand it correctly here we use reference count as a criteria of liveness. It might be good to add more explanations to the comment.

Also note that the same can be achieved using Guard pattern. See ReplicationConnGuard as an example.

Am I right that we will send callmemaybe requests forever until walproposer disconnects? Is it correct?

File: walkeeper/src/callmemaybe.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        match rx.recv_timeout(timeout) {
+            Ok(streaminfo) => {
+                info!("callmemaybe. thread_main. request_callback for timelineid={} tenantid={} refcount {}",
+                streaminfo.timelineid, streaminfo.tenantid, Arc::strong_count(&streaminfo));
+
+                let pageserver_addr = conf.pageserver_addr.as_ref().unwrap().clone();
+                let listen_pg_addr = conf.listen_pg_addr.clone();
+                let password = conf.pageserver_auth_token.clone();
+
+                runtime.block_on(async {
```
Comment (LizardWizzard) [inline walkeeper/src/callmemaybe.rs:76]: It seems that this is equivalent to using sync version. Do we need async here? 

If we use async there it becomes possible to run multiple callmemaybe calls concurrently. 

In current implementation the following is possible: imagine that we have two callmemaybe calls in the queue and the first one timeouts or is stuck for any other reason, in that case second call must wait for the first one. This behaviour might be undesirable. Solution to this is to use runtime.spawn, or something like that ie LocalSet or FuturesUnordered. But also it makes possible to issue multiple concurrent calls for the same pageserver/tenant/timeline. It might be desirable to avoid that

File: walkeeper/src/callmemaybe.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    Ok(())
+}
+
+pub fn thread_main(conf: SafeKeeperConf, rx: Receiver<Arc<WalStreamInfo>>) -> Result<()> {
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    let default_timeout = Duration::from_secs(5);
```
Comment (LizardWizzard) [inline walkeeper/src/callmemaybe.rs:62]: Might be better to move to a top level constant

File: walkeeper/Cargo.toml [hunk 4]
```diff
@@ -31,6 +31,8 @@ walkdir = "2"
 serde = { version = "1.0", features = ["derive"] }
 hex = "0.4.3"
 const_format = "0.2.21"
+tokio-postgres = "0.7.5"
```
Comment (LizardWizzard) [inline walkeeper/Cargo.toml:34]: tokio-postgres lives in the same repo as rust-postgres which we've already forked. Is it hard to point it to our forked version for less confusion in the future?

**Review by petuhovskiy — APPROVED** [review]
Looks good, should help with callmemaybe spamming that we can have now.

I'm not very familiar with tokio and may have missed some points. One thing that caught my eye is `current_thread::Runtime` with `block_on`. It seems that `request_callback` calls block current thread and we can have some problems with callmemaybe requests blocking each other, especially in case of timeouts and similar problems.

File: walkeeper/src/callmemaybe.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    let timeout = conf.recall_period.unwrap_or(default_timeout);
+    let mut addresses = Vec::new();
+
+    loop {
+        match rx.recv_timeout(timeout) {
```
Comment (funbringer) [inline walkeeper/src/callmemaybe.rs:67]: Here's an idea: with an async channel (e.g. [`tokio::sync::mpsc`](https://docs.rs/tokio/1.14.0/tokio/sync/mpsc/index.html)) it's possible to perform wait-for-timeout and wait-for-subscription concurrently! A simplified sketch:

```rust
enum Event {
  CallMeMaybe(...),
  Timeout,
}

// Whoa, async magic!
let event = tokio::select! {
    x = rx.recv() => CallMeMaybe(x),
    _ = tokio::time::sleep(...) => Timeout,
};

match event {
  ...
}
```

Furthermore, each "callmemaybe" subscription could be represented by a struct containing not only the request string but also a "last successful call timestamp", allowing us to unconditionally loop through a list of subscriptions and call their `send_request_if_time_has_passed()` (not a real name ;) methods.

TLDR: I'd rewrite the whole `thread_main` as an async function.

---

* https://docs.rs/tokio/1.14.0/tokio/macro.select.html (see notes on cancellation safety)

File: walkeeper/src/callmemaybe.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            Ok(streaminfo) => {
+                info!("callmemaybe. thread_main. request_callback for timelineid={} tenantid={} refcount {}",
+                streaminfo.timelineid, streaminfo.tenantid, Arc::strong_count(&streaminfo));
```
Comment (funbringer) [inline walkeeper/src/callmemaybe.rs:70]: @LizardWizzard has proposed a brilliant idea: we could implement "unsubscribe" as a guard, i.e. a struct with a destructor. The destructor could send an "unsubscribe" event containing `TentantId` and whatnots via a [`Sender`](https://docs.rs/tokio/1.14.0/tokio/sync/mpsc/struct.Sender.html).

It's also worth mentioning that tokio's mpsc supports [blocking send](https://docs.rs/tokio/1.14.0/tokio/sync/mpsc/struct.Sender.html#method.blocking_send).

**Review by funbringer — COMMENTED** [review]
Thanks for the patch!

**Review by funbringer — CHANGES_REQUESTED** [review]
Just using the scary red banner to highlight the fact that there are some concerns :)

File: walkeeper/src/callmemaybe.rs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let ps_addr_split: Vec<&str> = ps_addr.split(':').collect();
+    let mut config = tokio_postgres::Config::new();
+
+    config
+        .host(ps_addr_split[0])
+        .port(ps_addr_split[1].parse().unwrap());
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply walkeeper/src/callmemaybe.rs:28]: Maybe SocketAddr from std comes in handy

File: walkeeper/src/callmemaybe.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+async fn request_callback(
+    ps_addr: String,
+    listen_pg_addr: String,
+    pageserver_auth_token: Option<String>,
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let ps_addr_split: Vec<&str> = ps_addr.split(':').collect();
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs]: rewrote to use SocketAddr.
BTW, is there any reason we don't use it for all `*_addr` in SafeKeeperConf instead of String?

File: walkeeper/src/receive_wal.rs (hunk 1 above)
Comment (lubennikovaav) [reply walkeeper/src/receive_wal.rs]: Thank you for the idea. Rewrote to use the guard.

File: walkeeper/src/callmemaybe.rs (hunk 3 above)
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs:85]: I'd rather get rid of this entirely and introduce default `recall_period` because I don't see a use case for safekeepers with `recall` disabled.
@petuhovskiy , @arssher , what do you think?

File: walkeeper/src/callmemaybe.rs (hunk 5 above)
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs]: Thank you for the idea with `select!`. I rewrote the main to be async.

As for "last successful call timestamp" I am not so sure.  Do we need to be that accurate? I don't see any harm in sending a few extra calls.

File: walkeeper/src/callmemaybe.rs (hunk 3 above)
Comment (arssher) [reply walkeeper/src/callmemaybe.rs:85]: Makes sense.

File: walkeeper/src/callmemaybe.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                            streaminfo.timelineid, streaminfo.tenantid, e
+                        );
+                    });
+                });
+
+                if conf.recall_period.is_some() {
+                    addresses.push(streaminfo.clone());
+                }
+            }
+            Err(RecvTimeoutError::Timeout) => {
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs]: Indeed.
I rewrote it so that every incoming subscription is just added to the list, and then we call all addresses in the list.

File: walkeeper/src/callmemaybe.rs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    Ok(())
+}
+
+pub fn thread_main(conf: SafeKeeperConf, rx: Receiver<CallmeEvent>) -> Result<()> {
+    let runtime = runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(async { main_loop(conf, rx).await })
```
Comment (funbringer) [inline walkeeper/src/callmemaybe.rs:75]: I bet just `main_loop(conf, rx)` would do it.

File: walkeeper/src/callmemaybe.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply walkeeper/src/callmemaybe.rs]: I think nothing keeps us from doing that. I remember that in pageserver there are two config structures and one of them is used for toml serialization, so it may be necessary to add some custom handlers to parse socket addr from string. But worth checking. For safekeeper there is nothing similar because there is no persistent config, so we can just do it :)

File: walkeeper/src/receive_wal.rs [hunk 11]
```diff
@@ -128,19 +83,27 @@ impl<'pg> ReceiveWalConn<'pg> {
...
+        // as long as this wal_stream is alive, callmemaybe thread
+        // will send requests to pageserver
+        let _guard = match swh.conf.pageserver_addr {
+            Some(_) => {
+                // Need to establish replication channel with page server.
+                // Add far as replication in postgres is initiated by receiver, we should use callme mechanism
+                let timelineid = swh.timeline.get().timelineid;
+                let tx_clone = swh.tx.clone();
+                swh.tx
+                    .blocking_send(CallmeEvent::Subscribe(tenant_id, timelineid))
```
Comment (LizardWizzard) [inline walkeeper/src/receive_wal.rs:95]: You can use unbounded channel and then no blocking is needed

File: walkeeper/src/callmemaybe.rs [hunk 12]
```diff
@@ -0,0 +1,293 @@
+//!
```
Comment (LizardWizzard) [inline walkeeper/src/callmemaybe.rs:1]: Maybe it is a good idea to have top level docs mentioning what messages are sent when so the whole picture is covered

**Review by LizardWizzard — COMMENTED** [review]
Thanks! This is quite nice :) I've left a few suggestions here and there

File: walkeeper/src/callmemaybe.rs [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    let default_timeout = Duration::from_secs(5);
+    let recall_period = conf.recall_period.unwrap_or(default_timeout);
+
+    let subscriptions: Mutex<HashMap<(ZTenantId, ZTimelineId), SubscriptionState>> =
+        Mutex::new(HashMap::new());
+
+    loop {
+        let call_iteration = tokio::select! {
+            request = rx.recv() => {
+                match request {
```
[the remark this answers is not in the document]
Comment (funbringer) [reply walkeeper/src/callmemaybe.rs:228]: > Hm, thinking about how to reduce indentiation level.
 
+1 to decreasing indent. Frankly, I'd use question mark here, i.e. `rx.recv().ok_or_else(|| anyhow!("done!"))?` and handle the error somewhere near `thread_main`.

File: walkeeper/src/replication.rs [hunk 14]
```diff
@@ -221,6 +248,27 @@ impl ReplicationConn {
         };
         info!("Start replication from {:?} till {:?}", start_pos, stop_pos);
 
+        // Don't spam pageserver with callmemaybe queries
+        // when connection is already established.
```
Comment (arssher) [inline walkeeper/src/replication.rs:252]: slightly better to do that only when connection is indeed from pageserver, i.e. appname is not wal_proposer_recovery, see several lines above

File: walkeeper/src/callmemaybe.rs [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+            request_callback(
+                pageserver_addr,
+                listen_pg_addr,
+                password,
+                timelineid,
+                tenantid,
+            )
+            .await
+            .unwrap_or_else(|e| {
+                error!(
```
Comment (arssher) [inline walkeeper/src/callmemaybe.rs:193]: Seems like this error is never logged, better do that.

**Review by arssher — APPROVED** [review]

File: walkeeper/src/callmemaybe.rs [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+struct SubscriptionStateInner {
+    handle: Option<task::JoinHandle<()>>,
+    last_call_time: Instant,
+    paused: bool,
+}
+
+struct SubscriptionState {
+    tenantid: ZTenantId,
+    timelineid: ZTimelineId,
+    inner: Mutex<SubscriptionStateInner>,
```
[the remark this answers is not in the document]
Comment (arssher) [reply walkeeper/src/callmemaybe.rs:100]: current_thread runtime is used here (and we don't need more than that), so there is nothing to block/being blocked by. It would be nice to drop the mutex altogether actually.

[the remark this answers is not in the document]
Comment (LizardWizzard) [reply walkeeper/src/callmemaybe.rs:100]: If we hold the mutex across await points it might still make sense in a single threaded runtime. To actually spawn non send futures see [LocalSet](https://docs.rs/tokio/1.14.0/tokio/task/struct.LocalSet.html). Mutex might be replaced with Rc<RefCell>. But still beware of holding RefMut across await points, compiler should issue a warning in that case.

File: walkeeper/src/receive_wal.rs (hunk 11 above)
Comment (lubennikovaav) [reply walkeeper/src/receive_wal.rs]: Why would we need it if we send the message from a sync code anyway?

File: walkeeper/src/callmemaybe.rs [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+async fn request_callback(
+    ps_addr_str: String,
+    listen_pg_addr_str: String,
+    pageserver_auth_token: Option<String>,
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let mut config = tokio_postgres::Config::new();
+    let ps_addr = ps_addr_str.to_socket_addrs().unwrap().next().unwrap();
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs]: this code is gone after rebase

File: walkeeper/src/replication.rs (hunk 14 above)
Comment (lubennikovaav) [reply walkeeper/src/replication.rs]: Fixed. Do we need any similar handling in `ReceiveWalConn::run()`?

File: walkeeper/src/callmemaybe.rs [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+                            CallmeEvent::Unsubscribe(tenantid, timelineid) => {
+                                let mut subscriptions = subscriptions.lock().unwrap();
+                                subscriptions.remove(&(tenantid, timelineid));
+                                debug!("callmemaybe. thread_main. unsubscribe callback request for timelineid={} tenantid={}",
+                                timelineid, tenantid);
+                                false
+                            },
+                            CallmeEvent::Pause(tenantid, timelineid) => {
+                                let subscriptions = subscriptions.lock().unwrap();
+                                if let Some(sub) = subscriptions.get(&(tenantid, timelineid))
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply walkeeper/src/callmemaybe.rs]: it turned out that clippy doesn't like this style
https://rust-lang.github.io/rust-clippy/master/index.html#option_map_unit_fn

File: walkeeper/src/replication.rs (hunk 14 above)
Comment (arssher) [reply walkeeper/src/replication.rs]: No, it can only receive WAL from the proposer.

File: walkeeper/src/callmemaybe.rs (hunk 15 above)
Comment (arssher) [reply walkeeper/src/callmemaybe.rs]: uh, I was wrong and missed `error!` and `bail!` -- `error!` in the previous version was ok, but in the current code result of `request_callback` is unwrapped,  it will panic on failure.

File: walkeeper/src/callmemaybe.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply walkeeper/src/callmemaybe.rs]: That’s strange, and example doesn’t make sense to me. @funbringer what do you think about disabling this lint globally?

File: walkeeper/src/receive_wal.rs [hunk 19]
```diff
@@ -136,20 +90,33 @@ impl<'pg> ReceiveWalConn<'pg> {
...
-                .name("request_callback thread".into())
-                .spawn(move || {
-                    request_callback(conf, pageserver_connstr, timelineid, tenant_id);
+        // if requested, ask pageserver to fetch wal from us
+        // as long as this wal_stream is alive, callmemaybe thread
+        // will send requests to pageserver
+        let _guard = match self.pageserver_connstr {
+            Some(ref pageserver_connstr) => {
+                // Need to establish replication channel with page server.
+                // Add far as replication in postgres is initiated by receiver, we should use callme mechanism
```
Comment (LizardWizzard) [inline walkeeper/src/receive_wal.rs:99]: typo

File: walkeeper/src/receive_wal.rs [hunk 20]
```diff
@@ -136,20 +90,33 @@ impl<'pg> ReceiveWalConn<'pg> {
...
+                let timelineid = swh.timeline.get().timelineid;
+                let tx_clone = swh.tx.clone();
+                let pageserver_connstr = pageserver_connstr.to_owned();
+                swh.tx
+                    .send(CallmeEvent::Subscribe(
+                        tenant_id,
+                        timelineid,
+                        pageserver_connstr,
+                    ))
+                    .unwrap();
```
Comment (LizardWizzard) [inline walkeeper/src/receive_wal.rs:109]: It can fail when channel was closed or receiver was dropped, thats unlikely but lets add some context via expect or turn this to an error

File: walkeeper/src/replication.rs [hunk 21]
```diff
@@ -221,6 +248,28 @@ impl ReplicationConn {
         };
         info!("Start replication from {:?} till {:?}", start_pos, stop_pos);
 
+        // Don't spam pageserver with callmemaybe queries
+        // when connection to compute node is already established.
```
Comment (LizardWizzard) [inline walkeeper/src/replication.rs:252]: Comment is a bit misleading to me, we suspend callmemaybe polling when pageserver connects to us (safekeeper) and the polling will be continued if/when replication connection was destroyed

**Review by LizardWizzard — APPROVED** [review]
LGTM

Left few comments on minor leftovers

File: walkeeper/src/callmemaybe.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (funbringer) [reply walkeeper/src/callmemaybe.rs]: Personally, I agree with clippy (I wouldn't go as far as to impose my vision though). `map` better suits functional style, but returning unit `()` kind of implies that the statement is effectful (i.e. has side effects, or why else would you return unit?), and mixing both functional and effectful styles might lead to bugs. TLDR `if let` makes side effects stand out more.
